dpdispatcher/ssh_context.py

```diff
@@ -135,17 +135,48 @@ def _setup_ssh(self):
         ts.start_client(timeout=self.timeout)
 
         #Begin authentication; note that the username and callback are passed
-        key_path = os.path.join(os.path.expanduser("~"), ".ssh", "id_rsa")
-        if self.key_filename:
-            key_path = os.path.abspath(self.key_filename)
         key = None
         key_ok = False
         key_error = None
-        if os.path.exists(key_path):
-            try:
-                key = paramiko.RSAKey.from_private_key_file(key_path)
-            except paramiko.PasswordRequiredException:
-                key = paramiko.RSAKey.from_private_key_file(key_path, self.passphrase)
+        keyfiles = []
+        if self.key_filename:
+            key_path = os.path.abspath(self.key_filename)
+            if os.path.exists(key_path):
+                for pkey_class in (
+                    paramiko.RSAKey, 
+                    paramiko.DSSKey, 
+                    paramiko.ECDSAKey, 
+                    paramiko.Ed25519Key
+                ):
+                    try:
+                        # passing empty passphrase would not raise error.
+                        key = pkey_class.from_private_key_file(key_path, self.passphrase)
+                    except paramiko.SSHException as e:
+                        pass
+                    if key is not None:
+                        break
+        else:
+            for keytype, name in [
+                (paramiko.RSAKey, "rsa"),
+                (paramiko.DSSKey, "dsa"),
+                (paramiko.ECDSAKey, "ecdsa"),
+                (paramiko.Ed25519Key, "ed25519"),
+            ]:
+                for directory in [".ssh", "ssh"]:
+                    full_path = os.path.expanduser(
+                        "~/{}/id_{}".format(directory, name)
```

Comment:
This breaks the compatibility on the Windows platform. Use os.path.join or Pathlib instead.

Reply:
`os.path.expanduser` is just a Windows compatible implementation while add `ssh` in loop is an official implementation from `paramiko`.

Reply:
This is my result on Windows:
```py
PS C:\Program Files\Microsoft Visual Studio\2022\Community> python.exe
Python 3.10.8 (tags/v3.10.8:aaaf517, Oct 11 2022, 16:50:30) [MSC v.1933 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.path.expanduser("~/.ssh/id_rsa")
'C:\\Users\\njzjz/.ssh/id_rsa'
```

The correct output should be `'C:\\Users\\njzjz\\.ssh\\id_rsa'`.

Reply:
OK, I understand. Using `pathlib` might be a better choice.